src/basic/virt.c

```diff
@@ -227,12 +227,32 @@ static int detect_vm_xen(void) {
         return VIRTUALIZATION_XEN;
 }
 
+#define XENFEAT_dom0 11 /* xen/include/public/features.h */
+#define PATH_FEATURES "/sys/hypervisor/properties/features"
 /* Returns -errno, or 0 for domU, or 1 for dom0 */
 static int detect_vm_xen_dom0(void) {
         _cleanup_free_ char *domcap = NULL;
         char *cap, *i;
         int r;
 
+        r = read_one_line_file(PATH_FEATURES, &domcap);
+        if (r < 0 && r != -ENOENT)
+                return r;
+	if (r == 0) {
+                unsigned long features;
+
+                r = sscanf(domcap, "%lx", &features);
```

Comment:
i think safe_atolu() would be nicer here

Reply:
Thanks, I was looking for helpers like that, but missed them. Then I found many users of sscanf and used it as well.

Reply:
I thought about `safe_atolu`, but this rejects trailing whitespace. I don't have a XEN system around to test if it's present, but I think it'd be safest to explicitly truncate the string using `strstrip`.